codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -69,409 +71,1060 @@
+                    pipeline().fireChannelReadComplete();
+                }
+                return flushPending;
+            } finally {
+                inFireChannelReadComplete = false;
+                flushPending = false;
+            }
+        }
+
+        private final class Http2ChannelUnsafe implements Unsafe {
+            private final VoidChannelPromise unsafeVoidPromise =
+                    new VoidChannelPromise(DefaultHttp2StreamChannel.this, false);
+            @SuppressWarnings("deprecation")
+            private RecvByteBufAllocator.ExtendedHandle recvHandle;
+            private boolean writeDoneAndNoFlush;
+            private ChannelPromise pendingClosePromise;
+
+            @Override
+            public void connect(final SocketAddress remoteAddress,
+                                SocketAddress localAddress, final ChannelPromise promise) {
+                promise.setFailure(new UnsupportedOperationException());
+            }
+
+            @Override
+            public RecvByteBufAllocator.ExtendedHandle recvBufAllocHandle() {
+                if (recvHandle == null) {
+                    recvHandle = (RecvByteBufAllocator.ExtendedHandle) config().getRecvByteBufAllocator().newHandle();
+                }
+                return recvHandle;
+            }
+
+            @Override
+            public SocketAddress localAddress() {
+                return parent().unsafe().localAddress();
+            }
+
+            @Override
+            public SocketAddress remoteAddress() {
+                return parent().unsafe().remoteAddress();
+            }
+
+            @Override
+            public void register(EventLoop eventLoop, ChannelPromise promise) {
+                if (registered) {
+                    throw new UnsupportedOperationException("Re-register is not supported");
+                }
+
+                registered = true;
+
+                // Add the handler to the pipeline now that we are registered.
+                pipeline().addLast(inboundStreamHandler);
+
+                promise.setSuccess();
+
+                pipeline().fireChannelRegistered();
+                if (isActive()) {
+                    pipeline().fireChannelActive();
+                }
+            }
+
+            @Override
+            public void bind(SocketAddress localAddress, ChannelPromise promise) {
+                promise.setFailure(new UnsupportedOperationException());
+            }
+
+            @Override
+            public void disconnect(ChannelPromise promise) {
+                close(promise);
+            }
+
+            @Override
+            public void close(ChannelPromise promise) {
+                if (closePromise.isDone()) {
+                    promise.setFailure(new ClosedChannelException());
+                    return;
+                }
+                if (pendingClosePromise != null) {
+                    pendingClosePromise.addListener(new ChannelPromiseNotifier(promise));
```

Comment:
`new ChannelPromiseNotifier(promise)` -> `new DelegatingChannelPromiseNotifier(promise)`. This will remove the array and iteration behind the scenes.